src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java

```diff
@@ -63,70 +70,151 @@ public void dispose() {
         }
-
-        // return whatever we got (which may be null)
-        return (ServiceType) reference.service;
+        if (!result.isEmpty()) {
+            ServiceType[] srv = (ServiceType[]) Array.newInstance(type, result.size());
+            return result.toArray(srv);
+        }
+        return null;
     }
 
     /**
      * @see org.osgi.framework.ServiceListener#serviceChanged(org.osgi.framework.ServiceEvent)
      */
     public void serviceChanged(ServiceEvent event) {
-        final String[] objectClasses = (String[])event.getServiceReference().getProperty(Constants.OBJECTCLASS);
-        if ( objectClasses != null) {
-            for(final String key : objectClasses) {
-                Reference ref = null;
-                synchronized ( this ) {
-                    ref = this.cache.remove(key);
+        ServiceReference<?> serviceReference = event.getServiceReference();
+        final String[] objectClasses = (String[]) serviceReference.getProperty(Constants.OBJECTCLASS);
+        if (objectClasses != null) {
+            for (final String key : objectClasses) {
+                SortedSet<Reference> references;
+                synchronized (this) {
+                    references = this.cache.remove(key);
+                    if (references != null) {
+                        for (Reference reference : references) {
+                            bundleContext.ungetService(reference.getServiceReference());
+                        }
+                    }
                 }
-                if ( ref != null && ref != NULL_REFERENCE ) {
-                    this.bundleContext.ungetService(ref.reference);
+            }
+        }
+    }
+
+    private <ServiceType> SortedSet<Reference> getCachedReferences(Class<ServiceType> type) {
+        String key = type.getName();
+        SortedSet<Reference> references = cache.get(key);
+        if (references == null) {
+            references = Collections.synchronizedSortedSet(new TreeSet<>((o1, o2) -> -1 * o1.compareTo(o2)));
+            try {
+                Collection<ServiceReference<ServiceType>> serviceReferences = this.bundleContext.getServiceReferences(type, null);
+                if (!serviceReferences.isEmpty()) {
+                    List<ServiceReference<ServiceType>> localReferences = new ArrayList<>(serviceReferences);
+                    Collections.sort(localReferences);
+                    Collections.reverse(localReferences);
+                    for (ServiceReference<ServiceType> ref : localReferences) {
+                        references.add(new Reference(ref));
+                    }
+                    synchronized (this) {
+                        SortedSet<Reference> existing = this.cache.get(key);
+                        if (existing != null) {
+                            existing.addAll(
+                                    references.stream().filter(reference -> !existing.contains(reference)).collect(Collectors.toList()));
```

Comment:
I'm not sure the extra filter is needed - are we not already adding to a set?

Reply:
+1

Reply:
Got rid of this in [712b011](https://github.com/apache/sling-org-apache-sling-scripting-core/pull/13/commits/712b0115040cf239d59eeb8baad13a175ff17309).